sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -571,12 +657,535 @@ public boolean getValidate() {
+                counts.get(0), counts.get(1)));
+        throw new RuntimeException(errorMessage);
+      }
+
+      long filesCount = counts.get(0);
+      List<String> tempFiles = Lists.newArrayList();
+      IOChannelFactory factory = IOChannelUtils.getFactory(extractDestinationDir);
+      for (int i = 0; i < filesCount; ++i) {
+        String filePath =
+            factory.resolve(extractDestinationDir, String.format("%012d%s", i, ".avro"));
+        tempFiles.add(filePath);
+      }
+      return ImmutableList.copyOf(tempFiles);
+    }
+
+    private List<BoundedSource<TableRow>> createSources(
+        List<String> files, TableSchema tableSchema) throws IOException, InterruptedException {
+      final String jsonSchema = JSON_FACTORY.toString(tableSchema);
+
+      SerializableFunction<GenericRecord, TableRow> function =
+          new SerializableFunction<GenericRecord, TableRow>() {
+            @Override
+            public TableRow apply(GenericRecord input) {
+              try {
+                return AvroUtils.convertGenericRecordToTableRow(
+                    input, JSON_FACTORY.fromString(jsonSchema, TableSchema.class));
+              } catch (IOException e) {
+                throw new RuntimeException("Failed to convert GenericRecord to TableRow", e);
+              }
+            }};
+
+      List<BoundedSource<TableRow>> avroSources = Lists.newArrayList();
+      BackOff backoff = new AttemptBoundedExponentialBackOff(
+          MAX_FILES_VERIFY_ATTEMPTS, INITIAL_FILES_VERIFY_BACKOFF_MILLIS);
+      for (String fileName : files) {
+        while (BackOffUtils.next(Sleeper.DEFAULT, backoff)) {
+          if (IOChannelUtils.getFactory(fileName).getSizeBytes(fileName) != -1) {
+            break;
+          }
+        }
+        avroSources.add(new TransformingSource<>(
+            AvroSource.from(fileName), function, getDefaultOutputCoder()));
+      }
+      return ImmutableList.copyOf(avroSources);
+    }
+  }
+
+  private static class BigQueryReader extends BoundedSource.BoundedReader<TableRow> {
+    private final BigQuerySourceBase source;
+    private final BigQueryServices.BigQueryJsonReader reader;
+
+    private BigQueryReader(BigQuerySourceBase source, BigQueryServices.BigQueryJsonReader reader) {
+      this.source = source;
+      this.reader = reader;
+    }
+
+    @Override
+    public BoundedSource<TableRow> getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      return reader.start();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return reader.advance();
+    }
+
+    @Override
+    public TableRow getCurrent() throws NoSuchElementException {
+      return reader.getCurrent();
+    }
+
+    @Override
+    public void close() throws IOException {
+      reader.close();
+    }
+  }
+
+  @VisibleForTesting
+  static class TransformingSource<T, V> extends BoundedSource<V> {
```

Comment:
Special-casing it won't reduce the code, it can at most hide the SerializableFunction inside the "TransfromingSource". However, it will make this class harder to test.
(Cannot just using any source to test, have to build a AvroSource<GenericRecord>)

I don' see much benefits of special-casing the code.